File: benchmarks/cpp/string parsing/string_parsing.cpp
#include <iostream>
#include <string>
#include <chrono>

using namespace std;
using namespace chrono;

int main() {
    const int REPEAT = 100000; // change this to scale workload
    string pattern =
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        "abcd"; // total length = 100 chars

    string text;
    text.reserve(pattern.size() * REPEAT);

    for (int i = 0; i < REPEAT; i++) {
        text += pattern;
    }

    volatile long long count = 0;

    auto start = high_resolution_clock::now();

    for (char c : text) {
        if ((c >= 'a' && c <= 'z') ||
            (c >= 'A' && c <= 'Z')) {
            count++;
        }
    }

    auto end = high_resolution_clock::now();
    duration<double> elapsed = end - start;

    cout << "C++ String Parse Time: "
         << elapsed.count() << " seconds\n";

    return 0;
}

File: benchmarks/rust/string parsing/string_parsing.rs
use std::time::Instant;
use std::hint::black_box;

fn main() {
    let repeat: usize = 1_000_0000; // 100 MB
    let pattern =
        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789\
         abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789\
         abcd"; // 100 chars

    let mut text = String::with_capacity(pattern.len() * repeat);

    for _ in 0..repeat {
        text.push_str(pattern);
    }

    let mut count: u64 = 0;

    let start = Instant::now();

    let bytes = text.as_bytes();
    for i in 0..bytes.len() {
        let c = bytes[i];
        if (c >= b'a' && c <= b'z') ||
           (c >= b'A' && c <= b'Z') {
            count += 1;
        }
    }

    black_box(count); // prevents optimization

    let elapsed = start.elapsed();
    println!(
        "Rust String Parse Time: {:.6} seconds",
        elapsed.as_secs_f64()
    );
    println!("Count: {}", count);
}
